source/contribute.rst

```diff
@@ -41,128 +41,177 @@ contributing back to a Doctrine project.
+    for additional background information
 
-Topic Branches
---------------
 
-Topic branches are for contributors to develop bug fixes, new features,
-etc. so that they can be easily merged to **2.11.x**. They must follow a
-few rules as listed below:
+Time might have elapsed since the last time you contributed or since you
+cloned the repository. You might want to fetch the latest changes from
+the ``doctrine`` remote:
 
--  May branch off from: **2.11.x** whenever possible, or a newer version
-   branch otherwise. Keep in mind that your changes will be
-   up-merged to higher version branches by maintainers after the merge if
-   they are applicable.
--  Branch naming convention: anything except master, the default branch name,
-   or version branch names.
+.. code-block:: console
 
-Topic branches are used to develop new features and fix reported issues.
-When starting development of a feature, the target release in which this
-feature will be incorporated may well be unknown. The essence of a topic
-branch is that it exists as long as the feature is in development, but
-will eventually be merged into **2.11.x** or a release branch (to
-add the new feature or bugfix to a next release) or discarded (in case
-of a disappointing experiment).
+    $ git fetch doctrine
 
-Topic branches should exist in your **local** and **origin**
-repositories only, there is no need for them to exist in **doctrine**.
+.. note::
 
-Creating a topic branch
------------------------
+   Here, we assume ``doctrine`` points to the upstream repository, and
+   not your fork. If you are not sure, you can check with ``git remote -v``.
+
+I know it sounds awful, but the next step is to deal with one of the 2
+hard problems in computer science and come up with a name for your
+branch. Pick something meaningful.
 
-First create an appropriately named branch. When starting work on a new
-topic, branch off from **doctrine/2.11.x** or a **doctrine/\*** branch:
+If you have a feature to contribute that adds support for a new database
+called YourSQL, you could create a branch called ``your-sql-support``
+from the next minor branch:
 
 .. code-block:: console
 
-    $ git checkout -b fix-weird-bug doctrine/2.11.x
-    Switched to a new branch "fix-weird-bug"
+    $ git switch --create your-sql-support doctrine/2.4.x # use -c for short
 
-Now do some work, make some changes then commit them:
+
+Here, using ``doctrine/2.4.x`` instead of just ``2.4.x`` means you do
+not have to switch to 2.4.x and update it first. In fact, unless you are
+a maintainer, you are never supposed to commit on that branch. One way
+to avoid that mistake is to delete that branch:
+
+.. code-block:: console
+
+    $ git branch --delete 2.4.x # use -d for short
+
+Later, if you want to see how things are on ``doctrine/2.4.x``, you can
+use ``--detach``:
+
+.. code-block:: console
+
+    $ git switch --detach doctrine/2.4.x # use -d for short
+
```

Comment:
Starting from "In fact..." – do we really want to suggest this, and is it relevant in the context of creating a feature branch and committing on this?

Yes, it's an additional safeguard, but nothing you need to repeat every time you create a new branch.

Reply:
Let's remove it, it makes the guide longer.